File: src_fortran/sub/gapcalc.f
c
c
c
c
c
c
      subroutine GAPCALC(i)
c
c     determine GAP for one event
c
      implicit none
      integer i
      include '../inc/vel_com.inc'
      integer nofgaps,j,ig
      real xstn,ystn,xhyp,yhyp,dxstnhyp,dystnhyp
c
      integer iga(200)     !  max. 200 obs. for a single event
c
c     event: i  knobs(i)  +/-e(2,i)=x    e(3,i)=y    x=+/-x(nsta,1)  y=...,2)
c     station# are stored in array  ISTM(iobs,ievent)
c
c---- compute GAP for this event
      if(knobs(i).gt.200)then
         i f (.not.single_turbo)then
         write(16,*)' WARNING:'
         write(16,*)' Event# ',i,'   Nobs = ',knobs(i),' > 200 !!!'
         write(16,*)' Array IGA(100) is too small; redimension it in'
         write(16,*)' subr. GAPCALC'
         e n d i f
         write(6,*)' Event# ',i,'   Nobs = ',knobs(i),' > 200 !!!'
         write(6,*)' Array IGA(200) is too small; redimension it in'
         write(6,*)' subr. GAPCALC'
         stop'subr. GAPCALC >>> array IGA is too small !'
      endif
      nofgaps=0
      do j=1,knobs(i)
         if(w(j,i).gt.0.0)then
            nofgaps=nofgaps+1
            xstn=x(istm(j,i),1)
            ystn=x(istm(j,i),2)
            xhyp=e(2,i)
            yhyp=e(3,i)
cek    avoiding call to atan2(zero,zero)
cek              write(6,*) ' Gapcalc: i,nofgaps,xstn,ystn,xhyp,yhyp'
cek              write(6,'(1x,i4,2x,i3,2x,4f10.3)') i,nofgaps,
cek     +                  xstn,ystn,xhyp,yhyp
            dxstnhyp=abs(xstn-xhyp)
            dystnhyp=abs(ystn-yhyp)
            if(dxstnhyp.gt.0.0001.or.dystnhyp.gt.0.0001) then
              iga(nofgaps)=57.296*ATAN2(xstn-xhyp,ystn-yhyp)
            else
              iga(nofgaps)=359
            endif
cek
            if(iga(nofgaps).lt.0) iga(nofgaps)=iga(nofgaps)+360
         endif
      enddo
      if(nofgaps.gt.0)then
         call SORTI(iga,nofgaps)
      else
         write(6,*)'WARNING: Event-# :',i,'has zero observations!'
         if(.not.single_turbo)then
            write(16,*)'WARNING: Event-# :',i,'has zero observations!'
         endif
      endif
      igap(i)=0
      ig=iga(1)-iga(nofgaps)
      if(ig.lt.0) ig=ig+360
      if(ig.gt.igap(i)) igap(i)=ig
      do j=2,nofgaps
         ig=iga(j)-iga(j-1)
         if(ig.lt.0) ig=ig+360
         if(ig.gt.igap(i)) igap(i)=ig
      enddo
c---- IGAP(i) is the gap of this event (nr. i)
c
      return
      end ! of subr. gapcalc

File: src/sub/gapcalc.c
/*
 * Calculate azimuthal gap for earthquake location
 * Determines largest gap between stations around hypocenter
 * from Fortran GAPCALC subroutine
 */

#include <stddef.h>
#include <stdio.h>
#include <math.h>
#include <stdbool.h>
#include "../include/globals.h"

/* External global variables from vel_com.inc */
extern int knobs[IEQ];
extern bool single_turbo;
extern double w[IST][IEQ];
extern int istm[IST][IEQ];
extern double e[5][IEQ];
extern double x[IST][3];
extern int igap[IEQ];

/* External function */
extern void sorti(int arr[], int n);

void gapcalc(int i) {
    int nofgaps, j, ig;
    double xstn, ystn, xhyp, yhyp, dxstnhyp, dystnhyp;
    int iga[MAXOBSPEREVENT];
    
    nofgaps = 0;
    
    if (i < 0 || i >= IEQ) {
        return;
    }

    /* Check array size */
    if (knobs[i] > MAXOBSPEREVENT) {
        if (!single_turbo) {
            fprintf(stderr, "WARNING: Event# %d Nobs = %d > %d\n",
                    i, knobs[i], MAXOBSPEREVENT);
        }
        printf("Event# %d Nobs = %d > %d\n", i, knobs[i], MAXOBSPEREVENT);
        return;
    }
    
    /* Loop through observations and compute bearings */
    for (j = 0; j < knobs[i]; j++) {
        if (w[j][i] > 0.0) {
            nofgaps++;
            xstn = x[istm[j][i]][0];
            ystn = x[istm[j][i]][1];
            xhyp = e[1][i];  /* X coordinate of event */
            yhyp = e[2][i];  /* Y coordinate of event */
            
            dxstnhyp = fabs(xstn - xhyp);
            dystnhyp = fabs(ystn - yhyp);
            
            if (dxstnhyp > 0.0001f || dystnhyp > 0.0001f) {
                /* Compute bearing (degrees from north, clockwise) */
                iga[nofgaps - 1] = (int)(57.296f * atan2f(xstn - xhyp, ystn - yhyp));
            } else {
                iga[nofgaps - 1] = 359;
            }
            
            if (iga[nofgaps - 1] < 0) {
                iga[nofgaps - 1] = iga[nofgaps - 1] + 360;
            }
        }
    }
    
    if (nofgaps > 0) {
        sorti(iga, nofgaps);
    } else {
        printf("WARNING: Event-# %d has zero observations!\n", i);
        if (!single_turbo) {
            fprintf(stderr, "WARNING: Event-# %d has zero observations!\n", i);
        }
        return;
    }
    
    /* Calculate maximum gap */
    igap[i] = 0;
    ig = iga[0] - iga[nofgaps - 1];
    if (ig < 0) ig = ig + 360;
    if (ig > igap[i]) igap[i] = ig;
    
    for (j = 1; j < nofgaps; j++) {
        ig = iga[j] - iga[j - 1];
        if (ig < 0) ig = ig + 360;
        if (ig > igap[i]) igap[i] = ig;
    }
}
